unstructured/ingest/connector/hubspot.py

```diff
@@ -0,0 +1,365 @@
+
+from unstructured.ingest.error import SourceConnectionError
+from unstructured.ingest.interfaces import (
+    BaseConnectorConfig,
+    BaseIngestDoc,
+    BaseSessionHandle,
+    BaseSourceConnector,
+    ConfigSessionHandleMixin,
+    IngestDocCleanupMixin,
+    IngestDocSessionHandleMixin,
+    SourceConnectorCleanupMixin,
+    SourceMetadata,
+)
+from unstructured.ingest.logger import logger
+from unstructured.utils import requires_dependencies
+
+if t.TYPE_CHECKING:
+    from hubspot import HubSpot
+
+CONTENT_TAG = "content"
+
+
+class HubSpotObjectTypes(Enum):
+    CALLS = "calls"
+    COMMUNICATIONS = "communications"
+    EMAILS = "emails"
+    NOTES = "notes"
+    PRODUCTS = "products"
+    TICKETS = "tickets"
+
+
+@dataclass
+class HubSpotSessionHandle(BaseSessionHandle):
+    service: "HubSpot"
+
+
+@dataclass
+class SimpleHubSpotConfig(ConfigSessionHandleMixin, BaseConnectorConfig):
+    api_token: str = field(repr=False)
+    params: t.Optional[str] = None
+    properties: t.Optional[dict] = None
+    object_types: t.Optional[t.List[str]] = None
+    custom_properties: t.Optional[t.List[t.List[str]]] = None
+    mapped_custom_properties: t.Optional[t.Dict[str, t.List[str]]] = None
+
+    @requires_dependencies(["hubspot-api-client"], extras="hubspot")
+    def create_session_handle(self) -> HubSpotSessionHandle:
+        from hubspot import HubSpot
+
+        service = HubSpot(access_token=self.api_token)
+        return HubSpotSessionHandle(service=service)
+
+    def __post_init__(self):
+        if self.custom_properties is None:
+            return
+        for otype, cprop in self.custom_properties:
+            self.mapped_custom_properties[otype] = self.mapped_custom_properties.get(  # type: ignore
+                otype,
+                [],
+            ) + [
+                cprop,
+            ]
+
+
+@dataclass
+class HubSpotIngestDoc(IngestDocSessionHandleMixin, IngestDocCleanupMixin, BaseIngestDoc):
+    connector_config: SimpleHubSpotConfig
+    object_id: str
+
+    @property
+    def filename(self):
+        return (
+            Path(self.read_config.download_dir)
+            / f"{self.registry_name}/{self.object_id}.txt"  # type: ignore
+        ).resolve()
+
+    @property
+    def _output_filename(self):
+        return (
+            Path(self.processor_config.output_dir)
+            / f"{self.registry_name}/{self.object_id}.json"  # type: ignore
+        ).resolve()
+
+    @property
+    def record_locator(self) -> t.Optional[t.Dict[str, t.Any]]:
+        return {
+            "object_id": self.object_id,
+        }
+
+    @property
+    def version(self) -> t.Optional[str]:
+        return None
+
+    @property
+    def source_url(self) -> t.Optional[str]:
+        return None
+
+    def _fetch_obj(self, get_by_id_method, not_found_exception, **kwargs):
+        try:
+            response = get_by_id_method(self.object_id, **kwargs)
+        except not_found_exception as e:
+            logger.error(e)
```

Comment:
Do we want this to throw the `not_found_exception` here and let the pipeline logic either raise the error or simply log it there? @ryannikolaidis opinions?

Reply:
Yep, I think letting it raise makes sense. Not sure that hiding here brings value.

Reply:
the point of checking exists is to see if it exists on the remote service. this is ideally just an ls of sorts, not actually relying on fetching content. not possible in this case?

Reply:
...really that goes for any metadata. ideally doesn't rely on actually fetching the content

Reply:
If we need to make a call, would be worth seeing it there's a `HEAD` version since that doesn't require pulling down the data just to check for existence. I had to do that in the notion connector.

Reply:
that makes sense to me

Reply:
Can we omit the try-except and wrap this method with the `SourceConnectionNetworkError.wrap` decorator?